Index: htop/src/CRT.c
===================================================================
--- htop.orig/src/CRT.c
+++ htop/src/CRT.c
@@ -16,7 +16,6 @@ in the source distribution for its full
 #include <stdlib.h>
 #include <string.h>
 #include <locale.h>
-#include <langinfo.h>
 
 #define ColorPair(i,j) COLOR_PAIR((7-i)*8+j)
 
@@ -609,10 +608,7 @@ void CRT_init(int delay, int colorScheme
    setlocale(LC_CTYPE, "");
 
 #ifdef HAVE_LIBNCURSESW
-   if(strcmp(nl_langinfo(CODESET), "UTF-8") == 0)
-      CRT_utf8 = true;
-   else
-      CRT_utf8 = false;
+  CRT_utf8 = true;
 #endif
 
    CRT_treeStr =
